cpp/src/arrow/builder.cc

```diff
@@ -656,6 +658,206 @@ Status BooleanBuilder::Append(
+  return HashUtil::Hash(&value, sizeof(Scalar), 0);
+}
+
+template <typename T, typename Scalar>
+bool DictionaryBuilder<T, Scalar>::SlotDifferent(hash_slot_t index, const Scalar& value) {
+  const Scalar other = GetDictionaryValue(static_cast<int64_t>(index));
+  return other != value;
+}
+
+template <typename T, typename Scalar>
+Status DictionaryBuilder<T, Scalar>::AppendDictionary(const Scalar& value) {
+  return dict_builder_.Append(value);
+}
+
+#define BINARY_DICTIONARY_SPECIALIZATIONS(Type)                                       \
+  template <>                                                                         \
+  WrappedBinary DictionaryBuilder<Type, WrappedBinary>::GetDictionaryValue(           \
+      int64_t index) {                                                                \
+    int32_t v_len;                                                                    \
+    const uint8_t* v = dict_builder_.GetValue(static_cast<int64_t>(index), &v_len);   \
+    return WrappedBinary(v, v_len);                                                   \
+  }                                                                                   \
+                                                                                      \
+  template <>                                                                         \
+  int DictionaryBuilder<Type, WrappedBinary>::HashValue(const WrappedBinary& value) { \
+    return HashUtil::Hash(value.ptr_, value.length_, 0);                              \
+  }                                                                                   \
+                                                                                      \
+  template <>                                                                         \
+  bool DictionaryBuilder<Type, WrappedBinary>::SlotDifferent(                         \
+      hash_slot_t index, const WrappedBinary& value) {                                \
+    int32_t other_length;                                                             \
+    const uint8_t* other_value =                                                      \
+        dict_builder_.GetValue(static_cast<int64_t>(index), &other_length);           \
+    return !(other_length == value.length_ &&                                         \
+             0 == memcmp(other_value, value.ptr_, value.length_));                    \
+  }                                                                                   \
+                                                                                      \
+  template <>                                                                         \
+  Status DictionaryBuilder<Type, WrappedBinary>::AppendDictionary(                    \
+      const WrappedBinary& value) {                                                   \
+    return dict_builder_.Append(value.ptr_, value.length_);                           \
+  }
```

Comment:
I think this is OK on a first pass; we may want to return soon-ish and add some benchmarks on large-ish data. I suspect we may need to "flatten" the structure so that we aren't delegating to the child dictionary builder, but we will see